test/e2e/storage/csi_mock_volume.go

```diff
@@ -1165,6 +1169,118 @@ var _ = utils.SIGDescribe("CSI mock volume", func() {
+					}
+				})
+
+				_, claim, pod := createPod(false)
+				// Wait for PVC, Pod to get bound to make sure the CSI driver is fully started.
+				err := e2epv.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, f.ClientSet, f.Namespace.Name, claim.Name, time.Second, framework.ClaimProvisionTimeout)
+				framework.ExpectNoError(err, "while waiting for PVC to get provisioned")
+				err = e2epod.WaitForPodNameRunningInNamespace(m.cs, pod.Name, pod.Namespace)
+				framework.ExpectNoError(err, "failed to start pod")
+				err = e2epod.DeletePodWithWait(m.cs, pod)
+				framework.ExpectNoError(err, "failed to delete pod")
+				claim, err = m.cs.CoreV1().PersistentVolumeClaims(claim.Namespace).Get(ctx, claim.Name, metav1.GetOptions{})
+				framework.ExpectNoError(err, "failed to get pvc")
+
+				// Channel used to block test
+				stopCh := make(chan bool)
+
+				go func() {
+					defer ginkgo.GinkgoRecover()
+					// Create snapshot in a separate goroutine.
+					ginkgo.By("Creating snapshot")
+					_, snapshot := testsuites.CreateSnapshot(sDriver, m.config, testpatterns.DynamicSnapshotDelete, claim.Name, claim.Namespace)
+					framework.ExpectNoError(err, "failed to create snapshot")
+					test.snapshotNames = append(test.snapshotNames, snapshot.GetName())
+					ginkgo.By(fmt.Sprintf("Wait for finalizer to be added to claim %s", claim.Name))
+					err = e2epv.WaitForFinalizerAdded(ctx, m.cs, claim.Name, claim.Namespace, pvcAsSourceProtectionFinalizer, 1*time.Millisecond, 1*time.Minute)
+					framework.ExpectNoError(err)
+
+					// Unblock main thread since pvc finalizer has been added
+					stopCh <- true
+
+					ginkgo.By("Wait for Snapshot to be ready")
+					err = testsuites.WaitForSnapshotReady(m.config.Framework.DynamicClient, snapshot.GetNamespace(), snapshot.GetName(), framework.Poll, framework.SnapshotCreateTimeout)
+					framework.ExpectNoError(err)
+				}()
+
+				// Block until finalizer is added to pvc
+				<-stopCh
+				err = e2epv.DeletePersistentVolumeClaim(m.cs, claim.Name, claim.Namespace)
+				framework.ExpectNoError(err, "failed to delete pvc")
+
+				ginkgo.By(fmt.Sprintf("Wait for PV %s to be deleted", claim.Spec.VolumeName))
+				err = e2epv.WaitForPersistentVolumeDeleted(m.cs, claim.Spec.VolumeName, framework.Poll, 5*time.Minute)
+
+				ginkgo.By("Waiting for expected CSI calls")
+				err = wait.PollImmediateUntil(time.Second, func() (done bool, err error) {
+					_, index, err := compareCSICalls(trackedCalls, test.expectedCall, m.cs, m.config.DriverNamespace.Name, driverPodName, driverContainerName)
+					if err != nil {
+						return true, fmt.Errorf("error waiting for expected CSI calls: %s", err)
+					}
+					if index == 0 {
+						// No CSI call received yet
+						return false, nil
+					}
+					if len(test.expectedCall) == index {
+						// all calls received
+						return true, nil
+					}
+					return false, nil
+				}, ctx.Done())
+				framework.ExpectNoError(err, "while waiting for all CSI calls")
+			})
+		}
```

Comment:
Snapshots should be deleted too.  I don't see a DeleteSnapshot call.

For deleting snapshot, can you add a regular way to delete VolumeSnapshot first and another test case to delete VolumeSnapshotContent first?

Reply:
Currently i have `DeleteAndWaitSnapshot` called in test cleanup on line 1217. Do you want to move that here?

Reply:
It's fine to keep that there.  I think you can add a negative test here to delete VolumeSnapshotContent.  Since they are bound, it should only add the deletionTimestamp on VolumeSnapshotContent but it will not be deleted.  Just try to retrieve the VolumesnapshotContent from API server again after the delete call and check that the deletionTimestamp is added.  Then it will be cleaned up on line 1217.